worker/worker.py

```diff
@@ -133,139 +216,262 @@ def setup_parameters(config_file):
+
+    # When using --use_all_cores False when --use_all_cores True
+    # is in effect, we reset concurrency to the default
+    # value.
+
+    # We first want to look at the option --use_all_cores on
+    # the command line, before replacing it by the value from
+    # the config file.
+    if (
+        options.use_all_cores is not None
+        and options.use_all_cores
+        and options.concurrency is None
+    ):
+        options.concurrency = max_cpu_count
+    if (
+        options.use_all_cores is not None
+        and not options.use_all_cores
+        and options.concurrency is None
+        and config.getint("parameters", "concurrency") >= max_cpu_count
+    ):
+        print("Resetting concurrency to the default value {}".format(default_cpu_count))
+        options.concurrency = default_cpu_count
+    if options.use_all_cores is None:
+        options.use_all_cores = config.getboolean("parameters", "use_all_cores")
+    if options.concurrency is None:
+        options.concurrency = config.getint("parameters", "concurrency")
+
+    if options.concurrency > max_cpu_count or (
+        not options.use_all_cores and options.concurrency >= max_cpu_count
+    ):
+        print(
+            (
+                "\nYou cannot have concurrency {} but at most:\n"
+                "  a) {} with '--use_all_cores True';\n"
+                "  b) {} otherwise.\n"
+                "Please use option a) only if your computer is very lightly loaded.\n"
+            ).format(
+                options.concurrency,
+                max_cpu_count,
+                max_cpu_count - 1,
+            )
+        )
+        return None
 
-    if cpu_count <= 0:
+    if options.concurrency <= 0:
         print("Not enough CPUs to run fishtest: set '--concurrency' to at least one")
         return None
 
-    options.concurrency = str(cpu_count)
+    # Step 6: determine credentials
+
+    username, password = get_credentials(config, options, args)
+
+    if username == "":
+        print("Invalid or missing credentials")
+        return None
+
+    options.username = username
+    options.password = password
 
-    # Step 5: write command line parameters to the config file.
+    # Step 7: write command line parameters to the config file.
     config.set("login", "username", options.username)
     config.set("login", "password", options.password)
     config.set("parameters", "protocol", options.protocol)
     config.set("parameters", "host", options.host)
-    config.set("parameters", "port", options.port)
-    config.set("parameters", "concurrency", options.concurrency)
-    config.set("parameters", "max_memory", options.max_memory)
-    config.set("parameters", "min_threads", options.min_threads)
-    config.set("parameters", "fleet", options.fleet)
-    config.set("parameters", "use_all_cores", options.use_all_cores)
+    config.set("parameters", "port", str(options.port))
+    if options.concurrency != max_cpu_count:
```

Comment:
In master wins `use_all_cores = True` when both `concurrency = 5` and `use_all_cores = True` are in configfile, with the proposed code instead wins `concurrency = 5`
With this code we have only one entry between `concurrency = 5` and `use_all_cores = True`. Setting `--use_all_cores = False` without concurrency will use the default number of cores like the proposed code:
```python
    if options.concurrency != max_cpu_count:
        config.set("parameters", "concurrency", str(options.concurrency))
        config.remove_option("parameters", "use_all_cores")
    else:
        assert options.use_all_cores
        config.set("parameters", "use_all_cores", str(options.use_all_cores))
        config.remove_option("parameters", "concurrency")
    config.set("parameters", "max_memory", str(options.max_memory))
    config.set("parameters", "min_threads", str(options.min_threads))
    config.set("parameters", "fleet", str(options.fleet))

    with open(config_file, "w") as f:
```

BTW I agree with you that, without the retro compatibility requirement, it was way better something like `--concurrency max`

Reply:
I did it intentionally this way (it is written in the announcement). It is much more logical and for the common use cases it makes no difference.

Reply:
My suggested change uses your logic but writes only one of the two alternative options in `fishtest.cfg` and not both. IMO this is confusing:
```
[login]
username = user00
password = user00

[parameters]
protocol = https
host = dfts-0.pigazzini.it
port = 443
max_memory = 16224
min_threads = 1
fleet = False
use_all_cores = True
concurrency = 5
```

Reply:
Alternative:
```python
    if options.concurrency != max_cpu_count:
        config.set("parameters", "concurrency", str(options.concurrency))
        options.use_all_cores = str(False)
    else:
        assert options.use_all_cores
```